翻译v3.4自定义构建纹理压缩处理 (#1934)
cocos-creator/creator-docs: en/editor/publish/custom-build-plugin.md

> 2. After the build extension is created, you will see the generation path of the plugin in the **Console**. Click on the path to open the build extension package in the file manager of the operating system.

Reviewer: ```suggestion
2. After the build extension is created, notice the generation path of the plugin in the **Console**. Click on the path to open the build extension package in the file manager of the operating system.
```

    ![console](./custom-project-build-template/console.png)

3. Before enabling the build extension, execute `npm install` in the directory to install some dependent **@types** modules to compile normally. The interface definition that comes with the editor has been generated under the **@types** folder in the root directory. **Developer -> Export.d.ts** from the menu bar of the editor shows the latest interface definitions.

> 4. Click **Extension -> Extension Manager** in the menu bar of the editor to open the **Extension Manager** panel. Then select the **Project**/**Global** tab in the **Extension Manager**, and click the **Refresh Icon** button to see the build extension you just added. Then click the **Enable** button on the right to run the plug-in normally.

Reviewer: ```suggestion
4. Click **Extension -> Extension Manager** in the menu bar of the editor to open the **Extension Manager** panel. Then select the **Project**/**Global** tab in the **Extension Manager**, and click the **Refresh Icon** button to see the build extension just added. Then click the **Enable** button on the right to run the plug-in normally.
```

> 6. If you need to modify the content of the build extension, directly modify the build extension package under the `extensions` directory, see the `readme.md` file in the build extension package directory for details. Then find the corresponding build extension in the **Extension Manager**, and click the **Reload** icon button. At this time, the extension in the editor will re-run with the latest code and files.

Reviewer: ```suggestion
6. To modify the content of the built extension, directly modify the build extension package under the `extensions` directory, review the `readme.md` file in the build extension package directory for details. Find the corresponding build extension in the **Extension Manager**, and click the **Reload** icon button. The extension in the editor will re-run with the latest code and files.
```

    ![reload](./custom-project-build-template/reload.png)

## Basic configuration process

> To extend the build function of the plug-in, you need to add the `builder` field to the `contributions` in `package.json`, and the relative path configuration of the corresponding module can be passed to the specified platform in the field.

Reviewer: ```suggestion
To extend the build function of the plug-in, add the `builder` field to the `contributions` in `package.json`, and the relative path configuration of the corresponding module can be passed to the specified platform in the field.
```

Example `package.json`:

```json
// package.json

{
    "contributions": {
        "builder": "./dist/builder"
    }
}
```

> > **Note**: the `builder` field specifies the `./dist/builder.js` start script is the compiled script, and the source file of the start script is located in `./source/builder.ts`, if you need to configure the start script, please change it in the source file.

Reviewer: ```suggestion
> **Note**: the `builder` field specifies the `./dist/builder.js` start script is the compiled script, and the source file of the start script is located in `./source/builder.ts`. To configure the start script, change it in the source file.
```

}
```

### Custom texture compression processing

The `assetHandler` path configuration specified in the **start script configuration** above allows external developers to register some asset handling functions to replace the engine's handler module when building partial assets. Currently only **texture compression** handler registration is available.

Creator provides its own compression tools to handle compressed texture assets at build time, but does not focus on image compression because it needs to be compatible with different user environments and usually the compression tools are chosen to work on most computers rather than the most efficient ones. Therefore, Creator has opened up a plug-in mechanism in v3.4, **which allows users to directly register compression processing functions for the corresponding texture assets, which will be called at the appropriate processing time when building**.

The specific steps are as follows:

1. In the start script, write the relative path of the `assetHandlers` module script:

    ```ts
    export const assetHandlers = './asset-handlers';
    ```

> 2. In the `assetHandlers` script module, we open up the `compressTextures` function, so developers can write the corresponding handler function directly in `compressTextures`, which will be called during the texture compression phase of the build.

Reviewer: ```suggestion
2. In the `assetHandlers` script module, the `compressTextures` function has been opened up for developers to write the corresponding handler function directly in `compressTextures`, which will be called during the texture compression phase of the build.
```

    The handler function takes the current array of remaining unprocessed texture compression tasks and removes them from the original array when processing is complete. Texture compression tasks that are not removed are considered unprocessed and are placed in the next corresponding processing function until all processing functions have been processed, and if there are still unprocessed texture compression tasks, they are placed back in the Creator's original texture compression process.

    When there are multiple plugins registered with texture compression handler functions, they are executed in the order in which the plugins are started. If the previous plugin processes all the texture compression tasks, the subsequent plugins registered with the handler functions will not receive the tasks.

    The code example is as follows:

    ```ts
    type ITextureCompressType =
        | 'jpg'
        | 'png'
        | 'webp'
        | 'pvrtc_4bits_rgb'
        | 'astc_12x12'; // See interface definition for detailed format
    interface ICompressTasks {

>         src: string; // source file address

Reviewer 2: ```suggestion
        src: string; // Source file address
```

>         dest: string; // address of the generated target file (default suffix is PNG, other types need to be changed manually)

Reviewer 2: ```suggestion
        dest: string; // Address of the generated target file (default suffix is PNG, other types need to be changed manually)
```

When the code running in the main process is modified, the plugin must be restarted and then the process to be updated must be refreshed (this will be optimized later to try to solve the code update problem with a single restart, but refreshing is still the most thorough reloading method). The main process currently does not have a more appropriate debugging method, you can use the command line to open the editor to view the main process code log to assist debugging:

```bash
// Mac
/Applications/CocosCreator/Creator/3.0.0/CocosCreator.app/Contents/MacOS/CocosCreator --project projectPath

// Windows
... \CocosCreator.exe --project projectPath
```

### Rendering Process (Build Panel)

The start script of the build extension plugin has some fields that are registered to the **Build** panel, such as the display configuration of `options`, the `panel` field, and the `panel` script itself, which is loaded and executed in the render process. The rendering process is actually the window's own execution process. Open the DevTools to debug the `dom` elements, styles, scripts, etc. on the **Build** panel.

> If you modify the code registered to the **Build** panel, just refresh the panel without restarting the plugin.

Reviewer: ```suggestion
If the code registered to the **Build** panel is modified, refresh the panel without restarting the plugin.
```

- **Open the DevTools for the rendering process of the Build panel**

  Click on the **Build** panel and press the shortcut **Ctrl + Shift + I** (Windows) or **Command + Option + I** (Mac) to open the DevTools for the **Build** panel.

- **How to reload (refresh) the panel**

>   Press **Ctrl/Command + R** after clicking the debug tool in the **Build** panel or the **Build Release** panel.

Reviewer 2: ```suggestion
  Press **Ctrl/Command + R** after clicking on the **Build** panel or the DevTool in the **Build** panel.
```

### Build Process (`hooks` Script)

The actual execution phase of the build is a separate worker process, ensuring that even if an abnormal crash occurs, it will not affect the normal use of other windows. The scripts defined in the `hooks` field of the start script are also loaded and executed in this separate worker process.

If only the script defined in the `hook` field is modified, the build process can be refreshed without restarting the plugin. To do this, press **Ctrl/Command + R** after opening the build DevTools, as in the **Build** panel above.

#### Opening the DevTools for the build process

This includes the following three ways:

1. Click the **Open Build DevTools** button at the top right of the build task window in the **Build** panel.

>     ![dev_tools](./custom-project-build-template/dev_tools.png)

Reviewer 2: 这个图裂了

2. Click **Developers -> Open Build DevTools** in the main editor menu to open it.

3. In any plugin code or in the console, execute the following code:

    ```ts
    Editor.Message.send('builder', 'open-devtools');
    ```

>     You can extend this message method to suit your needs. For example, you can catch errors in the code you write to build the plugin, and automatically open DevTools once there is an exception or something like that.

Reviewer: ```suggestion
    `Editor.Message` can be extended to suit additional needs. For example, developers can catch errors in the code written to build the plugin, and automatically open DevTools once there is an exception or something like that.
```
